### PR TITLE
Retry github fetch calls

### PR DESCRIPTION
**Pull request by fejta**
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/test-infra/828)

### CODE REVIEW

File: jenkins/bootstrap.py [hunk 1]
```diff
@@ -119,9 +120,20 @@ def checkout(repo, branch, pull):
...
+        try:
+            call([
+                git, 'fetch', '--tags', repository(repo), ref,
+            ])
+        except subprocess.CalledProcessError as cpe:
+            if not retries:
+                raise
+            if cpe.returncode != 128:
+                raise
+            logging.warning('git fetch failed: %s', traceback.format_exc())
```
Comment (rmmh) [inline jenkins/bootstrap.py]: use logging.exception instead, it includes the traceback automatically

File: jenkins/bootstrap.py [hunk 2]
```diff
@@ -119,9 +120,20 @@ def checkout(repo, branch, pull):
     call([git, 'init', repo])
     os.chdir(repo)
     # TODO(fejta): cache git calls
-    call([
-        git, 'fetch', '--tags', repository(repo), ref,
-    ])
+    retries = 3
+    while retries > 0:
```
Comment (rmmh) [inline jenkins/bootstrap.py]: this might be pretty as something like `for x in range(2, -1, -1)`

File: jenkins/bootstrap.py [hunk 3]
```diff
@@ -119,9 +120,20 @@ def checkout(repo, branch, pull):
...
+        retries -= 1
+        try:
+            call([
+                git, 'fetch', '--tags', repository(repo), ref,
+            ])
```
Comment (rmmh) [inline jenkins/bootstrap.py]: You need to break if this is successful, either here or in a try-except-ese block.

File: jenkins/bootstrap.py (hunk 2 above)
Comment (fejta) [inline jenkins/bootstrap.py]: Done

File: jenkins/bootstrap.py (hunk 3 above)
Comment (fejta) [inline jenkins/bootstrap.py]: Done

File: jenkins/bootstrap.py (hunk 1 above)
Comment (fejta) [inline jenkins/bootstrap.py]: Done

File: jenkins/bootstrap.py [hunk 4]
```diff
@@ -118,10 +119,20 @@ def checkout(repo, branch, pull):
...
-    ])
+    retries = 3
+    for attempt in range(retries):
+        try:
+            call([
+                git, 'fetch', '--tags', repository(repo), ref,
+            ])
+            break
+        except subprocess.CalledProcessError as cpe:
+            if not attempt:
```
Comment (rmmh) [inline jenkins/bootstrap.py]: This will always be true the first time around.

Comment (fejta) [inline jenkins/bootstrap.py]: Wow! Great catch and embarrassing for me! I added unit tests to cover this.

File: jenkins/bootstrap_test.py [hunk 5]
```diff
@@ -121,6 +122,22 @@ def testOutput(self):
 class CheckoutTest(unittest.TestCase):
     """Tests for checkout()."""
 
+    def testFetchRetries(self):
+        self.tries = 0
+        expected_attempts = 3
+        def SecondTimeCharm(cmd, *a, **kw):
```
Comment (rmmh) [inline jenkins/bootstrap_test.py]: ThirdTimeCharm ?

Comment (krzyzacy) [inline jenkins/bootstrap_test.py]: good one...

**Pull request merged**